### PR TITLE
Fix issue 17968 - Use typeid(Object) instead of typeid(T) to avoid pulling in unused symbols

### PR DESCRIPTION
**Pull request by schveiguy**
The root cause of this problem still isn't fixed, I'll open a new bug report to demonstrate the original issue. This is a workaround, but since it's generated code by the compiler, it shouldn't really affect anything (except allow linking).

Note that using `typeid(Object)` instead of `typeid(T)` on the tuple member is fine as long as it's really an `Object`. `TypeInfo_Class.getHash` *always* casts its argument to `Object` and calls `toHash` on it, no matter what the actual type is.

ping @MartinNowak this is the bug that prevents `iopipe` from using `io`. Thanks to @tgehr, @jacob-carlborg , @thewilsonator, @rainers, @RazvanN7 for helping me figure this out at dconf!

### CODE REVIEW

**Review by jacob-carlborg — COMMENTED** [review]
@wilzbach is there a wrapper script that takes care of exiting early on failure?

**Comment by jacob-carlborg** [comment]
@wilzbach I’m talking about `set -e`. Is that still required in the scripts? You were working on something similar if I recall correctly.

**Comment by wilzbach** [comment]
Yep, `set -euo pipefail` is set by default.

**Comment by schveiguy** [comment]
FYI I discovered some more details about this bug on the plane. Will update

**Comment by schveiguy** [comment]
OK, so something I didn't realize originally, and it explains a lot of the weird behavior we were seeing: the members are all considered `const`.

This is why `is(T : Object)` did not work. It also explains the bizarre extra TypeInfo symbol. We never demangled it fully, it's actually the `Typeinfo_Const` for the object! I have a super-reduced test case for that one now, will add the bug report soon.

The updated fix I think should work and is still better than the original (even if we fix the other bug), as it doesn't add the extra Typeinfo unnecessarily. At least this proves that my test case I added is properly being run ;)

**Comment by schveiguy** [comment]
woohoo, it's starting to pass now.

**Comment by schveiguy** [comment]
New bug report: https://issues.dlang.org/show_bug.cgi?id=18833

**Review by jacob-carlborg — APPROVED** [review]

**Comment by JinShil** [comment]
Forgive my ignorance on this, but I'm having a hard time understanding the fundamental problem in this issue.  Why is `TypeInfo` needed for this?  I'm currently undertaking an effort to replace runtime calls that require `TypeInfo` with templates, so `TypeInfo` no longer be required for many things.  Would that help resolve this issue?

**Comment by JinShil** [comment]
Please look into the Darwin failure.  If it's spurious just create a new git hash for this PR with a new push.  It may need some babysitting to ensure it stays green until its reviewed and merged.

**Comment by JinShil** [comment]
If this does not fix the root cause, what does it fix?  Is it just a temporary workaround to be backed out when the root cause is fixed.  I'm having trouble understanding this PR.  What specifically is the problem, not the symptom, but the problem?

**Comment by schveiguy** [comment]
> Please look into the Darwin failure. If it's spurious

It was spurious. It's a test to see if the GC throws an InvalidMemoryOperationError when you allocate during a destructor. No way that was affected ONLY on osx 32-bit.

> just create a new git hash for this PR

Note: as a member of druntime, you can click on the test, and click on the "deprecate" link, and the tester will retest the PR. You need to click on the login link at the top of the autotester screen.

I did that for this test.

**Comment by schveiguy** [comment]
> If this does not fix the root cause, what does it fix?

The root cause is outlined more clearly in the new bug report: https://issues.dlang.org/show_bug.cgi?id=18833

> Is it just a temporary workaround to be backed out when the root cause is fixed?

I would say no, because using `typeid(const(Object))` generates less TypeInfo requirements, and provides the same exact call to `obj.toHash` that the original did.

> What specifically is the problem, not the symptom, but the problem?

The problem is that the compiler requires the TypeInfo of `const(SomeClass)` in order to generate the `xtoHash` function for a struct that contains that class, and the compiler fails to generate the TypeInfo in certain circumstances. This is all outlined in the new bug report, please ask more questions there if you still don't grok it.

**Comment by schveiguy** [comment]
FYI, I was working personally with @jacob-carlborg on this at dconf, so there is a lot he knows that is not in this PR conversation.

**Comment by jacob-carlborg** [comment]
As far as I understand it's kind of a workaround. In the sense that if the proper typeinfo was generated there wouldn't be a problem. But I wouldn't say that the workaround is any more wrong than the existing code, only slightly more complicated.

File: src/dmd/clone.d
Comment (n8sh) [inline src/dmd/clone.d]: Would `is(T == class) || is(T == interface)` be better? It would avoid false positives for non-objects that implicitly convert to objects, such as:

https://run.dlang.io/is/xlZFQA
```d
struct S
{
    size_t toHash() const { return 3; }
    Object what;
    alias what this;
}

void main()
{
	S s;
    static assert(is(S : const(Object)));
    static assert(!is(S == class));
    assert(typeid(S).getHash(&s) == 3);
}
```

**Comment by n8sh** [comment]
When a field is an Object can't we write `h = h * 33 + (p.tupleof[i] is null ? 0 : p.tupleof[i].toHash())` and avoid `TypeInfo` altogether? Is there a benefit to doing that betterC-wise?

**Comment by schveiguy** [comment]
`xtoHash` I believe is only for the struct's TypeInfo anyway. I'm not sure exactly what happens on betterC when you don't specify toHash. AA's certainly don't work, so that's a moot point.

File: src/dmd/clone.d
Comment (schveiguy) [reply src/dmd/clone.d]: OK, what about this:

```D
is(T* : const(Object)*)
```

I tested it against types with alias this, and it seems to work.

Comment (jacob-carlborg) [reply src/dmd/clone.d]: `__traits(getLinkage)` works on classes and interfaces now https://github.com/dlang/dmd/commit/971777de5a19b960c343265efca3f9833ea071bc.

**Comment by schveiguy** [comment]
> avoid TypeInfo altogether?

Note that we would have to cast away const to do that (as Object.toHash is not const), and it is subject to alias this trickery as well.

It is a good idea, though...

**Comment by jacob-carlborg** [comment]
>> When a field is an Object can't we write h = h * 33 + (p.tupleof[i] is null ? 0 : p.tupleof[i].toHash()) and avoid TypeInfo altogether? Is there a benefit to doing that betterC-wise?

> Note that we would have to cast away const to do that (as Object.toHash is not const)

This is basically what `TypeInfo_Class.getHash` is doing. Casting away `const`, null check and calling `toHash`.

**Comment by schveiguy** [comment]
> This is basically what TypeInfo_Class.getHash is doing.

Not exactly -- it's casting it FROM const(void)* to Object directly. Then calling it. I think I'd have to go through the same set of casts in order to get it to work.

However, one thing I'd rather not do is put more complication into this generated function -- nobody sees the code anywhere. I'd rather have the breaking of const guarantees inside the runtime.

**Comment by schveiguy** [comment]
Alternatively, we can find a way to fix the root cause and avoid all this BS :)

File: src/dmd/clone.d
Comment (jacob-carlborg) [reply src/dmd/clone.d]: Can you please add the test case from n8sh.

Comment (schveiguy) [reply src/dmd/clone.d]: @jacob-carlborg ok

**Comment by n8sh** [comment]
By "fix the root cause" do you mean avoid using `TypeInfo` as @JinShil in https://github.com/dlang/dmd/pull/7799#issuecomment-365048319:

>That is not actually a problem with this PR, but a general problem with D's ubiquitous use of runtime type information (i.e. `TypeInfo`) for things that are known at compile-time. As I mentioned previously, I intend to champion an effort replace such uses of `TypeInfo` with templates, which should actually eliminate all such errors, and produce messages that are more meaningful.

**Comment by schveiguy** [comment]
@JinShil I think it's a good idea, but I don't want to make this PR into that change.

I'm going to add the @n8sh's case as a test case, but don't have time this moment.

**Comment by schveiguy** [comment]
@n8sh That's not what I meant, I mean fix issue 18833 -- the compiler should be generating typeinfo that is being used.

File: src/dmd/clone.d
Comment (MartinNowak) [inline src/dmd/clone.d]: OT: We should look at whether Knuth hashing is actually a sane choice here.
[Multiplicative hashing](https://en.wikipedia.org/wiki/Hash_function#Multiplicative_hashing)

File: src/dmd/clone.d
Comment (MartinNowak) [inline src/dmd/clone.d]: The need for comparing pointers is slightly subtle.

**Review by MartinNowak — APPROVED** [review]
Let's try this, it's a weird workaround though.

File: src/dmd/clone.d
Comment (jacob-carlborg) [reply src/dmd/clone.d]: Checking for a class and using `__traits(getLinkage)` to exclude C++ and Objective-C classes might have been more obvious.

File: src/dmd/clone.d
Comment (n8sh) [reply src/dmd/clone.d]: @MartinNowak I share your hope that this will be subject to further examination. I changed to this in April (https://github.com/dlang/dmd/pull/8175) because it was a small improvement that could pass quickly through the approval pipeline, not because it was the final word. In particular I'd like if we could use compile-time information about the types and order of the fields to make better choices.

File: src/dmd/clone.d
Comment (schveiguy) [reply src/dmd/clone.d]: Really, this is never seen by anyone. So I don't know that it matters what technique is used. I'm hoping that at some point instead of adjusting this, we simply move the function generation into an object.d template.

I'm also hoping that the underlying issue is fixed, and this workaround becomes moot.

**Pull request merged**
